### PR TITLE
 Improve GetDetectorOffsets's peak fitting algorithm

### LINKED ISSUE #22192: Improve GetDetectorsOffsets for better peak fitting
### Expected behavior

Recently it is found that a two-step Gaussian/PseudoVoigit peak fitting may be able to find the peak center more precisely than the current algorithm in GetDetectorOffsets for VULCAN data.  

The feature will be added to GetDetectorOffsets while its default behavior won't be changed.

### PR DESCRIPTION
**Pull request by wdzhou**
It is added as an option such that each peak can be fitted for second time with well defined peak range at FWHM in order to fit the peak center precisely.

An optional TableWorkspace can be output for fitted peak parameters for review.

**To test:**

1. The default behavior of GetDetectorOffsets shall be same as before.
2. All the unit tests, new and old, shall be passed.

Fixes #22192. 

**Release Notes** 

It is added to [diffraction release note](https://github.com/mantidproject/mantid/pull/22201/files#diff-b2bd2540c595f643521322a749fbd5dc)
<!--

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/inc/MantidAlgorithms/GetDetectorOffsets.h [hunk 1]
```diff
@@ -61,14 +73,27 @@ class DLLExport GetDetectorOffsets : public API::Algorithm {
   void init() override;
   void exec() override;
   /// Call Gaussian as a Child Algorithm to fit the peak in a spectrum
-  double fitSpectra(const int64_t s, bool isAbsolbute);
+  double fitSpectra(const int64_t s, bool isAbsolbute, const double xmin,
```
Comment (peterfpeterson) [inline Framework/Algorithms/inc/MantidAlgorithms/GetDetectorOffsets.h]: Rename to `fitPeakFirstTime` to make it easier to think about the code

File: Framework/Algorithms/src/GetDetectorOffsets.cpp [hunk 2]
```diff
@@ -164,21 +221,116 @@ void GetDetectorOffsets::exec() {
...
+ *  @param use_fit_result:
  *  @return The calculated offset value
  */
-double GetDetectorOffsets::fitSpectra(const int64_t s, bool isAbsolbute) {
-  // Find point of peak centre
-  const auto &yValues = inputW->y(s);
-  auto it = std::max_element(yValues.cbegin(), yValues.cend());
-  const double peakHeight = *it;
-  const double peakLoc = inputW->x(s)[it - yValues.begin()];
+double GetDetectorOffsets::fitSpectra(
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/GetDetectorOffsets.cpp]: The workspace index parameter should be renamed to `wi' or `index' or anything more direct than `s`.

File: Framework/Algorithms/src/GetDetectorOffsets.cpp [hunk 3]
```diff
@@ -107,16 +129,46 @@ void GetDetectorOffsets::exec() {
...
+                        << fit_result.height << ") or too-narrow peak ("
+                        << fit_result.sigma << ")\n";
+      }
+
+      offset = offset2;
+    }
+
     double mask = 0.0;
-    if (std::abs(offset) > m_maxOffset) {
+    if (mask_it || std::abs(offset) > m_maxOffset) {
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/GetDetectorOffsets.cpp]: Fold this into the logic of lines 154-168.

File: Framework/Algorithms/src/GetDetectorOffsets.cpp [hunk 4]
```diff
@@ -164,21 +221,116 @@ void GetDetectorOffsets::exec() {
...
+ * @return
+ */
+double GetDetectorOffsets::fitPeakSecondTime(
+    size_t wi, const bool isAbsolute, const double minimum_peak_height,
+    GetDetectorsOffset::PeakLinearFunction &fit_result, bool &mask_it) {
+  // check y
+  double offset(0);
+  mask_it = false;
+
+  if (minimum_peak_height != EMPTY_DBL() &&
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/GetDetectorOffsets.cpp]: `this->isEmpty()` is the correct check to use. Also, if the default value of the parameter was zero, this extra check could go away.

File: Framework/Algorithms/src/GetDetectorOffsets.cpp [hunk 5]
```diff
@@ -164,21 +221,116 @@ void GetDetectorOffsets::exec() {
...
+    g_log.warning() << "ws-index " << wi << " peak height " << fit_result.height
+                    << " too low!"
+                    << "\n";
+  } else if (fit_result.sigma < 0.5) {
+    // doesn't meet the minimum peak sigma
+    mask_it = true;
+  }
+
+  if (mask_it) {
+    offset = -1;
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/GetDetectorOffsets.cpp]: Why don't you set the offset to zero?

File: Framework/Algorithms/src/GetDetectorOffsets.cpp [hunk 6]
```diff
@@ -164,21 +221,116 @@ void GetDetectorOffsets::exec() {
...
+    return offset;
+  }
+
+  // also check whether there is any nan in vector Y
+  auto vec_y = inputW->histogram(wi).y();
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/GetDetectorOffsets.cpp]: Change to `const auto &points = inputW->histogram(wi).points();`

File: Framework/Algorithms/src/GetDetectorOffsets.cpp [hunk 7]
```diff
@@ -227,21 +404,38 @@ double GetDetectorOffsets::fitSpectra(const int64_t s, bool isAbsolbute) {
   return offset;
 }
 
-/**
- * Create a function string from the given parameters and the algorithm inputs
- * @param peakHeight :: The height of the peak
- * @param peakLoc :: The location of the peak
+//-----------------------------------------------------------------------------------------
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/GetDetectorOffsets.cpp]: If you are not going to fill in the doxygen, delete it.

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Overall it looks like a reasonable set of changes. Two things that need to be dealt with:
* `fitPeak...` overloads the return value as an error code and the result. This is a bad design that will be difficult to sort out later on.
* Don't be afraid to improve the existing code (e.g. rename variables and functions) to clarify what is going on in the larger picture.

File: Framework/Algorithms/src/GetDetectorOffsets.cpp [hunk 8]
```diff
@@ -72,6 +75,25 @@ void GetDetectorOffsets::init() {
...
+      "Peak fitting range for the first round fitting wil be determined by "
+      "given XMin and XMax"
+      "And in second round peak fitting, fit range will be limited to FHWM. ");
+
+  declareProperty("OutputFitResult", false,
+                  "If true, then a TableWorkspace containing all the fitted "
+                  "parameters' values will be output with the name as "
+                  "$OUTPUTWORKSPACE_FitResult");
+
+  declareProperty(
```
[the remark this answers is not in the document]
Comment (wdzhou) [reply Framework/Algorithms/src/GetDetectorOffsets.cpp]: cannot find disableWhen in any cpp file.

File: Framework/Algorithms/src/GetDetectorOffsets.cpp (hunk 3 above)
Comment (wdzhou) [reply Framework/Algorithms/src/GetDetectorOffsets.cpp]: change here is not trivial.  need to look twice.

File: Framework/Algorithms/src/GetDetectorOffsets.cpp (hunk 5 above)
Comment (wdzhou) [reply Framework/Algorithms/src/GetDetectorOffsets.cpp]: There is some logic error here!

File: Framework/Algorithms/src/GetDetectorOffsets.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (peterfpeterson) [reply Framework/Algorithms/src/GetDetectorOffsets.cpp]: My mistake, it is `EnabledWhenProperty ` ([example](https://github.com/mantidproject/mantid/blob/master/Framework/Algorithms/src/CalculateFlatBackground.cpp#L46)).

File: Framework/Algorithms/src/GetDetectorOffsets.cpp [hunk 9]
```diff
@@ -224,91 +235,139 @@ void GetDetectorOffsets::exec() {
...
-                << "; new fit range: " << xmin << ", " << xmax << "\n";
-  offset = fitSpectra(wi, isAbsolute, xmin, xmax, fit_result, true);
-
-  return offset;
+  return mask_it;
 }
 
 //-----------------------------------------------------------------------------------------
-/** Calls Gaussian1D as a child algorithm to fit the offset peak in a spectrum
+//** Fit the peak for the second time
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/GetDetectorOffsets.cpp]: Remove commented out code

File: Framework/Algorithms/src/GetDetectorOffsets.cpp [hunk 10]
```diff
@@ -407,11 +470,11 @@ double GetDetectorOffsets::fitSpectra(
 //-----------------------------------------------------------------------------------------
 /** Create a function string from the given parameters and the algorithm inputs
  * @brief GetDetectorOffsets::createFunction
- * @param peakHeight
- * @param peakLoc
- * @param peakSigma
- * @param a0
- * @param a1
+ * @param peakHeight : peak height
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/GetDetectorOffsets.cpp]: The names of the parameters are clear enough that that don't need to be documented.

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Delete some comments and you're good.

**Comment by NickDraper** [comment]
@wdzhou Are we aiming for this release with this one?  It has been stale, and failing since January.

**Comment by martyngigg** [comment]
Note that the target branch will now need changing for it to make 4.0

**Comment by NickDraper** [comment]
@wdzhou @peterfpeterson Could you decide between the two of you whether this branch should get some active time to progress it, or whether it should be closed (and the branch left open if it has future value).

**Comment by NickDraper** [comment]
No progress since Jan 23rd, pushing to the next release

**Comment by NickDraper** [comment]
This PR is being closed as it is over a year old and has not been recently progressed. If it is important to progress this work reopen the PR and prioritise the work.
Otherwise, the underlying branch will be removed at the next cleanup along with the next release unless it is progressed further.

**Pull request closed without merging**